Add date alerts to reasons

oliverguenther opened the pull request:
- [x]   The _reason_ provided for a date alert notification (comparable to &quot;mentioned&quot; or &quot;watched&quot;) will be &quot;Date alert&quot;.
- [x]   When the user is inside the sidebar element &quot;Date alert&quot; the 3rd line of information always display the related info with the date alert in the style specified in the visuals. As the {time} is relative, this will change accordingly with the current date:
    - [x]   &quot;Start date is {time}&quot; (&quot;Start date was {time}&quot; in case of past)
    - [x]   &quot;Finish date is {time}&quot;
    - [x]   Once the finish date is in the past, the work package is considered overdue: &quot;Overdue since {time}&quot;
    - [x]   {time} is the relative difference between the current date and the set date in question
- [x]   In case the user is in any of the other sidebar elements (eg. Inbox) there are two possibilities:
    - [x]   There is only one notification for the work package and it is a date alert: in this case the 3rd line will follow the same logic specified above.
    - [x]   There is one date alert but also other notifications concerning that work package: the date alert isn&#39;t immediately visible, we will display the normal information &quot;X days ago by Name Surname&quot;.
- [x]   In case start and finish date are on the same day and the configuration is also configured equally for both start and finish date, only display the finish date information as that one is more important.
- [x]   In the specific case of MILESTONES as they only have a single date, the finish date of the configuration will govern whether to create a notification.
    - [x]   Display text that is specific to milestones: &quot;Milestone date is {time}&quot;.
    - [x]   Overdue should the same for milestone-typed WPs as for any other type.
- [x]   The information needs to reflect changes done to the work package after the notification is created.
    - [x]   E.g. User is notified about the work package starting in 2 days. Another user moves the work package a week into the future. Then the notification card needs to say that it is starting in 9 days.
- [x] When a notification is emitted and the work package's due/start date is unset in the meantime, the frontend receives a detail of `null` and shows "is deleted" matching the activity text.


https://community.openproject.org/wp/43683

oliverguenther committed:
Add date alerts to reasons

https://community.openproject.org/wp/43683

oliverguenther committed:
Extract current authors footer into own component

oliverguenther committed:
Add date alert component

oliverguenther committed:
Add date alert deriving value from details

oliverguenther committed:
Allow implict-arrow-linebreak

oliverguenther committed:
Make property and/or overdue bold

oliverguenther committed:
Fix showing date alert in other filters

oliverguenther committed:
Add spec for date alerts displaying

oliverguenther committed:
Add spec to have the notification update in case the date updates

oliverguenther committed:
Find the interesting alert (use dueDate if multiple)

oliverguenther committed:
Add explicit ID desc sort to not confuse developers

oliverguenther committed:
Extend spec to check for double alerts

oliverguenther committed:
Adapt spec to now milestone specific text

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.html:
```diff
@@ -0,0 +1,27 @@
+    >
+    {{ text_for_additional_authors(actors.length - 3) }}
+  </span>
+  </div>
```
Mostly cosmetic, but there's a ton of weird indentation going on above. I also think this wrapping div might not be necessary?

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.sass:
```diff
@@ -0,0 +1,16 @@
+.op-ian-actors
+  display: grid
+  grid-template-columns: auto 1fr
+  grid-column-gap: 5px
```
Can we do this with spot spacings? 0.25rem would be 4px though @psatyal

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.sass:
```diff
@@ -0,0 +1,16 @@
+  &--date
+    @include text-shortener
+    max-width: 100%
+    line-height: 16px
```
Can also be a spot spacing

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.html:
```diff
@@ -0,0 +1,27 @@
+<div class="op-ian-actors">
```
You can ditch this div and set `@HostBinding('class.op-ian-actors')` instead. You'llprobably also need `ViewEncapsulation.None` then, but that's fine imo

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.ts:
```diff
@@ -0,0 +1,109 @@
+  text_for_additional_authors(number:number):string {
+    let hint:string;
+    if (number === 1) {
+      hint = this.text.and_other_singular;
```
if you return immediately you can get rid of the mutable variable `hint`

b12f commented on frontend/src/app/features/in-app-notifications/entry/actors-line/in-app-notification-actors-line.component.ts:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  isMobile():boolean {
+    return this.deviceService.isMobile;
```
I don't think you need the overhead of a function call here. I'd just check for `deviceService.isMobile` in the template directly

b12f commented on frontend/src/app/features/in-app-notifications/entry/date-alert/in-app-notification-date-alert.component.html:
```diff
@@ -0,0 +1,12 @@
+<div
+  class="op-ian-date-alert"
+  [ngClass]="{ 'op-ian-date-alert_overdue': isOverdue }"
+>
```
Here we can save a wrapping div also by using `HostBinding`

b12f commented on frontend/src/app/features/in-app-notifications/entry/in-app-notification-entry.component.ts:
```diff
@@ -44,34 +40,31 @@ export class InAppNotificationEntryComponent implements OnInit {
+    .pipe(
+      map((reason) => reason === 'date_alert'),
+      map((dateAlertFiltered) => {
+        const dateAlerts = this.aggregatedNotifications.filter((notification) => notification.reason === 'dateAlert');
```
we can save this calculation by returning dateAlertFiltered if available before

oliverguenther replied:
That's not the same. dateAlertFiltered just returns true if the reason has been filtered (date alert menu is active), but doesn't imply if any notifications are of type date alert

b12f reviewed (changes requested):
Mostly nitpicks

oliverguenther committed:
Use HostBinding and spot css

oliverguenther committed:
Handle null date values

oliverguenther committed:
Fix text if date is today

b12f reviewed (approved):
LGTM

cbliard committed:
Fix time dependent flaky test

When running on some hours of the day, like 23:15, today is not the same
for UTC, Paris, or Kathmandu time zones. When creating work packages
starting next day, we have to be explicit about in which time zone the
next day is to ensure the data set is correct.
